tools/kani-cov/src/report.rs

```diff
@@ -0,0 +1,262 @@
+
+    let checked_format = check_format(&args.format);
+
+    for file in source_files {
+        let fun_info = function_info_from_file(&file);
+        let mut file_cov_info = Vec::new();
+        for info in fun_info {
+            let cov_results = function_coverage_results(&info, &file, &results);
+            let line_coverage = line_coverage_results(&info, &cov_results);
+            let line_coverage_matched: Vec<(usize, Option<(u32, MarkerInfo)>)> =
+                (info.start.0..=info.end.0).zip(line_coverage.clone()).collect();
+            file_cov_info.push(line_coverage_matched);
+        }
+        output_coverage_results(&checked_format, file, file_cov_info)?;
+    }
+
+    Ok(())
+}
+
+/// Validate arguments to the `report` subcommand in addition to clap's
+/// validation.
+pub fn validate_report_args(_args: &ReportArgs) -> Result<()> {
+    // No validation is done at the moment
+    Ok(())
+}
+
+/// Checks the `format` argument for the `report` subcommand and selects to
+/// another format if the one we are checking is not possible for any reason.
+///
+/// For example, if the `Terminal` format is specified but we are not writing to
+/// a terminal, it will auto-select the `Escapes` format to avoid character
+/// issues.
+fn check_format(format: &ReportFormat) -> ReportFormat {
+    let is_terminal = std::io::stdout().is_terminal();
+    match format {
+        ReportFormat::Terminal => {
+            if is_terminal {
+                ReportFormat::Terminal
+            } else {
+                ReportFormat::Escapes
+            }
+        }
+        ReportFormat::Escapes => ReportFormat::Escapes,
+    }
+}
+
+/// Output coverage results, and highlight regions or lines which are not
+/// covered.
+///
+/// When highlighting the terminal, highlighted areas must be closed on the each
+/// line to avoid highlighting beyond the source code area. Therefore, this
+/// highlighting algorithm emits codes for both opening and closing the
+/// highlighted area if any highlighting must be done for a line.
+///
+/// Therefore, a coverage region spanning multiple lines will be represented
+/// with multiple highlighted lines. For example, let's take this uncovered
+/// function:
+///
+/// fn _other_function() {
+///     println!("Hello, world!");
+/// }
+///
+/// It will be highlighted as follows:
+///
+/// ```fn _other_function() {'''
+/// ```    println!("Hello, world!");'''
+/// ```}'''
+///
+/// Note how there is always a corresponding escape for each line.
+/// This is more sophisticated than the naive solution, which would simply emit two escapes
+/// per region (one opening, one closing) as follows:
+///
+/// ```fn _other_function() {
+///     println!("Hello, world!");
+/// }'''
+///
+pub fn output_coverage_results(
+    format: &ReportFormat,
+    filepath: PathBuf,
+    results: Vec<LineResults>,
+) -> Result<()> {
+    let flattened_results: Vec<(usize, Option<(u32, MarkerInfo)>)> =
```

Comment:
Couldn't this be empty?

Reply:
Yes, this could be empty. For that to happen, there must be no functions in the file. But if there's at least one, and even if it's unreachable, `flattened_results` will have been populated with full-line markers.